Habitica/src/main/java/com/habitrpg/android/habitica/interactors/UseCase.java

```diff
@@ -0,0 +1,34 @@
+package com.habitrpg.android.habitica.interactors;
+
+import com.habitrpg.android.habitica.executors.PostExecutionThread;
+import com.habitrpg.android.habitica.executors.ThreadExecutor;
+
+import rx.Observable;
+
+public abstract class UseCase<Q extends UseCase.RequestValues, T> {
+
+    private final ThreadExecutor threadExecutor;
+    private final PostExecutionThread postExecutionThread;
+
+    protected UseCase(ThreadExecutor threadExecutor,
+                      PostExecutionThread postExecutionThread) {
+        this.threadExecutor = threadExecutor;
+        this.postExecutionThread = postExecutionThread;
+    }
+
+    /**
+     * Builds an {@link rx.Observable} which will be used when executing the current {@link UseCase}.
+     */
+    protected abstract Observable<T> buildUseCaseObservable(Q requestValues);
+
+    @SuppressWarnings("unchecked")
+    public Observable<T> observable(Q requestValues) {
+        return this.buildUseCaseObservable(requestValues)
+                .subscribeOn(postExecutionThread.getScheduler())
```

Comment:
this causes all usecases to be run on the main thread (since that's the postexecutionthread we're using). In the clean-architecture branch, I did this because realm objects always have to be handled on the same thread and this was a working temporary solution. but we should probably find a way to at least have some of the usecases run on a background thread. Suggestions for how to implement this best are appreciated.